### ISSUE TITLE
Protocol Type Incompatibility: str not assignable to str | None despite being less strict

### ISSUE THREAD

**Issue by Filimoa** [microsoft/pyright#9595]
**Describe the bug**
In the following example, you will see that although the protocol is less strict than what's being passed (`str`), it raises an error that the type is not exactly `str | None`

**Code or Screenshots**
```python
from dataclasses import dataclass
from typing import Protocol

class SampleProtocol(Protocol):
    name: str | None

@dataclass
class SampleObj:
    name: str # more strict than the protocol

def print_age(obj: SampleProtocol):
    if obj.name:
        print(obj.name)
    else:
        print("No name")


var = SampleObj(name="John")
print_age(var) # Type "str" is not assignable to type "str | None"
```
print_age will have the following error:
```
Argument of type "SampleObj" cannot be assigned to parameter "obj" of type "SampleProtocol" in function "print_age"
  "SampleObj" is incompatible with protocol "SampleProtocol"
    "name" is invariant because it is mutable
    "name" is an incompatible type
      Type "str" is not assignable to type "str | None"
        "str" is not assignable to "None"  (reportArgumentType)
```

I thought maybe this is because the dataclass is mutable, but making it a frozen dataclass does not change anything. 

**VS Code extension or command-line**
Tried this in vscode and the [online playground](https://pyright-play.net/?pythonVersion=3.12&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sBYAKFEigQE8j8A7AcwONPIAK4BGGRhU9emkx4AynGKpsQsCLGoAFCrXiAlAC56UY1A4Ls%2BqBgQgoAHygA5MB2yS6AAXhJptOr7kFIiUAeQAjACtDOhNTc0trEHcYbGAoIhBOBAB9OC5sDTBIy3lFZWFRPWjY-DSiiIA6MwgLI1iTDKzCyKbzXTbjbFQcavbjTo4EDQAiZziW6f66dwA3OFsAXihS4OxwiI1m7A3pgCkwAAsORfoJnLyCtZBdIA)

**Comment by erictraut** [comment]
Pyright is working correctly here and in accordance with the typing spec, so this isn't a bug.

Your protocol indicates that the type must accept a value of type `str | None` for the `name` field. That means someone could attempt to write a `None` value to this field and still comply with the protocol. However, this would violate the `SampleObj` type. For example, consider if `print_age` included the statement `obj.name = None`.

Mutable attributes are necessarily treated as _invariant_ by a type checker. It sounds like you expect it to be treated as _covariant_, which would allow for the "less strict" behavior that you're looking for.

There is a draft [PEP 767](https://peps.python.org/pep-0767/) that allows an attribute for attributes to be marked `ReadOnly`. Read-only attributes can be safely treated as covariant. This PEP is still in the development phase and hasn't yet been implemented in pyright, but it is something I plan to [add soon](https://github.com/microsoft/pyright/issues/9559).
